core-plugins/src/main/java/co/cask/hydrator/plugin/batch/aggregator/RowDenormalizerAggregator.java

```diff
@@ -71,28 +77,24 @@ public void initialize(BatchRuntimeContext context) throws Exception {
   }
 
   @Override
-  public void groupBy(StructuredRecord record, Emitter<StructuredRecord> emitter) throws Exception {
-    Schema keyFieldSchema = Schema.recordOf("record", Schema.Field.of(keyField, Schema.of(Schema.Type.STRING)));
-    StructuredRecord.Builder builder = StructuredRecord.builder(keyFieldSchema);
-    builder.set(keyField, record.get(keyField));
-    emitter.emit(builder.build());
+  public void groupBy(StructuredRecord record, Emitter<String> emitter) throws Exception {
+    emitter.emit(record.get(keyField).toString());
```

Comment:
don't call toString(). You should be validating that the key field is actually a string (at which point you can just cast it), and that it is not null before emitting anything. I'm seeing this across a bunch of commits, but toString() is usually meant for display purposes. If you are using it in the logic, it is probably wrong.